be/src/env/env_hdfs.cpp

```diff
@@ -2,137 +2,80 @@
-    DCHECK(_opened);
-    size_t read_size = 0;
-    Status st = _read_at(offset, res->data, res->size, &read_size);
-    if (!st.ok()) return st;
-    res->size = read_size;
+    tSize n = static_cast<tSize>(std::min<uint64_t>(std::numeric_limits<tSize>::max(), res->size));
+    tSize r = hdfsPread(_fs, _file, offset, res->data, n);
+    if (UNLIKELY(r == -1)) {
+        return Status::IOError(fmt::format("fail to hdfsPread {}: {}", _file_name, get_hdfs_err_msg()));
+    }
+    res->size = r;
     return Status::OK();
 }
 
 Status HdfsRandomAccessFile::read_at(uint64_t offset, const Slice& res) const {
-    size_t read_size = 0;
-    RETURN_IF_ERROR(_read_at(offset, res.data, res.size, &read_size));
-    if (read_size != res.size) {
-        return Status::InternalError(
-                strings::Substitute("fail to read enough data, file=$0, offset=$1, size=$2, expect=$3", _file_name,
-                                    offset, read_size, res.size));
+    if (UNLIKELY(res.size > std::numeric_limits<tSize>::max())) {
+        return Status::NotSupported("read size is greater than std::numeric_limits<tSize>::max()");
+    }
+    tSize r = hdfsPreadFully(_fs, _file, offset, res.data, res.size);
+    if (UNLIKELY(r == -1)) {
+        return Status::IOError(fmt::format("fail to hdfsPreadFully {}: {}", _file_name, get_hdfs_err_msg()));
     }
     return Status::OK();
 }
 
-Status HdfsRandomAccessFile::readv_at(uint64_t offset, const Slice* res, size_t res_cnt) const {
-    // TODO: implement
-    return Status::InternalError("HdfsRandomAccessFile::readv_at not implement");
-}
-
-// ===================================  S3RandomAccessFile  =========================================
-
-S3RandomAccessFile::S3RandomAccessFile(S3Client* client, const std::string& bucket, const std::string& object,
-                                       size_t object_size)
-        : _client(client), _bucket(bucket), _object(object), _object_size(object_size) {
-    _file_name = "s3://" + _bucket + "/" + _object;
-}
-
-Status S3RandomAccessFile::read(uint64_t offset, Slice* res) const {
-    size_t read_size = 0;
-    Status st = _client->get_object_range(_bucket, _object, offset, res->size, res->data, &read_size);
-    if (!st.ok()) return st;
-    res->size = read_size;
-    return st;
-}
-
-Status S3RandomAccessFile::read_at(uint64_t offset, const Slice& res) const {
-    size_t read_size = 0;
-    RETURN_IF_ERROR(_client->get_object_range(_bucket, _object, offset, res.size, res.data, &read_size));
-    if (read_size != res.size) {
-        return Status::InternalError(
-                strings::Substitute("fail to read enough data, file=$0, offset=$1, size=$2, expect=$3", _file_name,
-                                    offset, read_size, res.size));
+Status HdfsRandomAccessFile::size(uint64_t* size) const {
+    if (_file_size == 0) {
+        auto info = hdfsGetPathInfo(_fs, _file_name.c_str());
```

Comment:
According to my knowledge, this function call crashes all the time. better to put `file_size` into constructor.

Reply:
> According to my knowledge, this function call crashes all the time. better to put `file_size` into constructor.

I wrote a test case to cover the method, and it works without crash

Reply:
Real access to hdfs? wierd.

Reply:
> Real access to hdfs? wierd.

Yes, real access to HDFS.